exo/inference/pytorch/model/hf.py

```diff
@@ -0,0 +1,107 @@
+import torch
+import torch.nn as nn
+import numpy as np
+
+from transformers import AutoModelForCausalLM, LlamaConfig, DynamicCache, Cache
+from exo.inference.shard import Shard
+from exo.helpers import DEBUG
+from typing import Tuple
+
+from .utils import sample_logits
+
+class ShardedHuggingFaceModel(torch.nn.Module):
+    def __init__(self, shard: Shard):
+        super(ShardedHuggingFaceModel, self).__init__()
+
+        if DEBUG >= 2:
+            print(f"\nShardedHuggingFaceModel init with shard {shard}")
+
+        self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+        self.shard = shard
+
+        
+
+        # Load the model
+        self.full_model = AutoModelForCausalLM.from_pretrained(
+            shard.model_id,
+            torch_dtype=torch.float16 if torch.cuda.is_available() else torch.float32,
+            device_map="auto"
+        )
+
+        # using llamaconfig not working setting layers manually
+        layers = []
+        for i in range(shard.start_layer, shard.end_layer + 1):
+            layer = self.full_model.model.layers[i]
+
+            if DEBUG >= 2:
+                print(f"Loading layers[{i}]")
+
+            layers.append(layer)
+        
+        self.full_model.model.layers = nn.ModuleList(layers)
+
+        if DEBUG >= 2:
+            print(f"full_model.model layer: {len(self.full_model.model.layers)}")
+
+        # Embeddings and final layer norm
+        # used for doing what forward LlamaModel does in transformers
+        self.embed_tokens = self.full_model.model.embed_tokens
+        self.norm = self.full_model.model.norm
+
+        # self.past_key_values = DynamicCache()
+
+    def forward_layers(
```

Comment:
I really like this approach of generalising this so it works for other models without having to explicitly implement them.

Can you write a test for a model with a different architecture to make sure this generalises e.g. recurrent Gemma?
I wonder if we need a little bit of model-specific behaviour to enable this in general?